fix #21426: Windows GC only wakes up one thread when doing parallel scanning

rainers opened the pull request:
use manual event reset to wake up all background scanning threads, because sys.core.event.setIfInitialized only wakes up one thread otherwise. These threads keep spinning until busyThreads becomes 0, which only happens after the event is manually reset from the main scanning thread.

rainers committed:
fix #21426: Windows GC only wakes up one thread when doing parallel scanning

use manual event reset to wake up all background scanning threads, because sys.core.event.setIfInitialized only wakes up one thread otherwise. These threads keep spinning until busyThreads becomes 0, which only happens after the event is manually reset from the main scanning thread.

rainers commented:
Too late/risky for stable?

rainers commented:
The `evDone` event is only used for a controlled shutdown of the GC. When used at the end of the scan with the scanStack running dry, `evDone.wait` is used as a "sleep" for a millisecond. I guess this can be handled more efficiently, but I'm not sure whether always signaling when empty/non-empty is more efficient in all situations, e.g. when a long chain of nodes in a list results in only few entries to be on the stack. A small amount of work might be handled more efficiently in a single thread.

Anyway, I think that is separate issue from threads not running at all.

schveiguy commented:
> The `evDone` event is only used for a controlled shutdown of the GC.

It's set by every background thread whenever the thread has exited the scan section.

https://github.com/dlang/dmd/blob/8762d1eaee42119269b82a1b1b7063c89c1e6a69/druntime/src/core/internal/gc/impl/conservative/gc.d#L3736

My understanding is the evDone is to get all the threads back to sleeping on the evStart event.

schveiguy commented:
I also am concerned about the way busyThreads are managed. I see a potential case where there is data left to scan, but a perfect storm happens and all the threads become unbusy, even though that data is still there, and the threads exit.

For instance, if you inserted a sleep [here](https://github.com/dlang/dmd/blob/8762d1eaee42119269b82a1b1b7063c89c1e6a69/druntime/src/core/internal/gc/impl/conservative/gc.d#L3783) then might it be more likely to expose this race?

The whole thing seems very wrong to me.

In any case, another issue for another time. Clearly it doesn't happen very much, or at all, as missing a scan of data could be disastrous for the GC.

schveiguy commented:
Note that an auto-reset event might actually be appropriate in that case! The thread that wakes up can re trigger if there's still data left.

rainers commented:
> I would think just one event is needed, and all it does is say "there is stuff ready to scan", and it's set when you add something to the list, and reset when you remove the last thing from the list. Just that event and the stopGC flag I think would be enough?

I implemented that, but results for short GC cycles get worse:
```
MIN conalloc         0.863 s,     5 MB, 3356 GC  354 ms, Pauses  350 ms <    0 ms
MIN conappend        0.202 s,     5 MB,  590 GC   72 ms, Pauses   70 ms <    0 ms
MIN concpu           0.883 s,     5 MB, 3193 GC  364 ms, Pauses  359 ms <    0 ms
MIN conmsg           1.357 s,    12 MB,   79 GC   17 ms, Pauses   13 ms <    0 ms
MIN dlist            1.128 s,    22 MB,   51 GC   91 ms, Pauses   89 ms <    2 ms
MIN huge_single      0.025 s,  1501 MB,    2 GC    0 ms, Pauses    0 ms <    0 ms
MIN rand_large       1.705 s,   165 MB, 6884 GC  593 ms, Pauses  281 ms <    0 ms
MIN rand_small       4.133 s,    22 MB, 3596 GC  555 ms, Pauses  230 ms <    0 ms
MIN slist            1.056 s,    35 MB,   19 GC   45 ms, Pauses   44 ms <    2 ms
MIN testgc3          1.508 s,   330 MB,   11 GC   45 ms, Pauses   42 ms <    9 ms
MIN tree1            0.334 s,    12 MB,   34 GC   33 ms, Pauses   31 ms <    1 ms
MIN tree2            0.727 s,     1 MB,  216 GC    5 ms, Pauses    4 ms <    0 ms
MIN vdparser         1.276 s,    70 MB,   13 GC   29 ms, Pauses   24 ms <    3 ms
MIN words            0.721 s,   396 MB,    5 GC   10 ms, Pauses    8 ms <    7 ms
```
The benchmarks are not very reproducable as can be seen by the different number of GC invocations, but the trend seems to be: tests that only run a few but longer GC cycles (testgc3, vdparser) benefit from the parallel scanning (16 threads in my case), but others with many but shorter collections take longer overall. I suspect that comes from the synchronization overhead.

> Note that an auto-reset event might actually be appropriate in that case! The thread that wakes up can re trigger if there's still data left.

Indeed, maybe too many threads are fighting over the next stack entry.

rainers committed:
GC: simplify background thread handling using manual event to signal non-empty scan stack

rainers commented:
Switched back to master, as it might not come without side-effects. I also pushed the simplified event handling.

rainers committed:
GC: switch back to auto-reset events, but allow another thread to run if the scanStack becomes non-empty or a new thread sees that the stack still has entries

rainers commented:
> > Note that an auto-reset event might actually be appropriate in that case! The thread that wakes up can re trigger if there's still data left.
> 
> Indeed, maybe too many threads are fighting over the next stack entry.

Implemented that, the results improve quite a bit:
```
MIN conalloc         0.775 s,     5 MB, 2963 GC  278 ms, Pauses  275 ms <    0 ms
MIN conappend        0.196 s,     5 MB,  758 GC   73 ms, Pauses   71 ms <    0 ms
MIN concpu           0.795 s,     5 MB, 3548 GC  305 ms, Pauses  302 ms <    0 ms
MIN conmsg           1.267 s,    12 MB,   58 GC   10 ms, Pauses    6 ms <    1 ms
MIN dlist            1.088 s,    22 MB,   51 GC   69 ms, Pauses   67 ms <    1 ms
MIN huge_single      0.024 s,  1501 MB,    2 GC    0 ms, Pauses    0 ms <    0 ms
MIN rand_large       1.578 s,   162 MB, 8737 GC  439 ms, Pauses  200 ms <    0 ms
MIN rand_small       3.284 s,    22 MB, 3596 GC  393 ms, Pauses  169 ms <    0 ms
MIN slist            1.075 s,    35 MB,   19 GC   48 ms, Pauses   47 ms <    3 ms
MIN testgc3          1.524 s,   330 MB,   11 GC   42 ms, Pauses   38 ms <    8 ms
MIN tree1            0.349 s,    12 MB,   85 GC   58 ms, Pauses   56 ms <    1 ms
MIN tree2            0.499 s,     1 MB,  216 GC    4 ms, Pauses    3 ms <    0 ms
MIN vdparser         1.253 s,    70 MB,   13 GC   26 ms, Pauses   22 ms <    2 ms
MIN words            0.722 s,   396 MB,    5 GC   10 ms, Pauses    9 ms <    7 ms
```
It doesn't seem to harm the short GC cycles anymore, but the longer ones benefit from parallelism.
BTW: do you have a more reliable set of benchmarks?

schveiguy commented:
Ohh! there is this loop: https://github.com/dlang/dmd/blob/8762d1eaee42119269b82a1b1b7063c89c1e6a69/druntime/benchmark/gcbench/conalloc.d#L39-L44

So this is going to be a huge wildcard. Why do we have this loop in there?

schveiguy commented:
Maybe up the N value to get more activity? Or limit the number of times this loop can run?

In any case, I don't think we can trust the total time numbers there as it will fluctuate with the number of times that loop can trigger a GC, but maybe dividing the time by the number of GC cycles is a reasonable measurement?

schveiguy commented on druntime/src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -3662,7 +3682,7 @@ Lmark:
 
-        evStart.initialize(false, false);
+        evStackFilled.initialize(false, false);
         evDone.initialize(false, false);
```
I think we can completely remove evDone. It's only waited on by the main thread with a 1ms timeout when the GC is shutting down. Could either replace with a sleep, or just remove this whole "exit loop" altogether -- if you set the stopGC flag, have the background threads set the evStackFilled event on exit, to wake up the next one.

Then the main thread will sleep on joining the background threads, and that is the true end of the loop.

rainers replied:
I thought so too, but wasn't sure whether that can fail and never terminate. IIRC there can be spurious wake-ups when waiting for events, maybe a wake-up might also be missed sometimes? Maybe an unconditional sleep is a simple and safe version.

rainers commented:
> Maybe up the N value to get more activity? Or limit the number of times this loop can run?
> 
> In any case, I don't think we can trust the total time numbers there as it will fluctuate with the number of times that loop can trigger a GC, but maybe dividing the time by the number of GC cycles is a reasonable measurement?

I think these tests are too CPU heavy and don't really test the GC. Increasing N will only let the execution time explode with only a fraction spent in the GC, so we should just ignore them for now.

I made the "ultimate" test, building all the phobos unittests with all files on a single command line. Here are the results with different GC options:
```
command              Process Mem / Time
dmd:                     19.4 GB   1:58.830 min:sec  GC memory:10187 MB,    1 GC  3212 ms, Pauses  1772 ms < 1772 ms
dmd -lowmem parallel:0:  12.3 GB   2:50.590 min:sec  GC memory: 7883 MB,   18 GC 46983 ms, Pauses 45594 ms < 7100 ms
dmd -lowmem parallel:1:  12.3 GB   2:31.680 min:sec  GC memory: 7883 MB,   18 GC 28956 ms, Pauses 27538 ms < 4231 ms
dmd -lowmem parallel:4:  12.3 GB   2:18.060 min:sec  GC memory: 7883 MB,   18 GC 17524 ms, Pauses 16081 ms < 2404 ms
dmd -lowmem parallel:7:  12.3 GB   2:17.120 min:sec  GC memory: 7883 MB,   18 GC 12620 ms, Pauses 11117 ms < 1585 ms
dmd -lowmem parallel:11: 12.3 GB   2:10.860 min:sec  GC memory: 7883 MB,   18 GC 11585 ms, Pauses 10079 ms < 1409 ms
dmd -lowmem parallel:15: 12.3 GB   2:12.030 min:sec  GC memory: 7883 MB,   18 GC 11766 ms, Pauses 10310 ms < 1433 ms
dmd -lowmem disable:1    20.0 GB   2:17.370 min:sec  GC memory:15435 MB,    1 GC  5564 ms, Pauses  4770 ms < 4770 ms
dmd -lowmem gc:precise   12.4 GB   2:21.620 min:sec  GC memory: 7883 MB,   18 GC 12570 ms, Pauses 11201 ms < 1563 ms
													 
dmd BumpGC:              30.0 GB   2:01.140 min:sec  
dmd cleanup:none         19.4 GB   1:55.280 min:sec  GC memory:10187 MB,    0 GC     0 ms, Pauses     0 ms <    0 ms
```

- this is run on a mobile AMD Ryzen 7 8845HS with 8 cores+HT and 64 GB RAM,. The latter causes the GC to run rather seldomly.
- measurements can vary by a couple of seconds
- parallel:1 should be similar to the currently released version
- GC and pause times scale rather well up to parallel:7 (8 threads), though a bit slower than 1/threads. After that there is not that much improvement, probably due to hyper-threading not being full cores.
- even though dmd is meant to use the bump allocator by default it still allocates half of its memory with the disabled GC. This is because the lowered druntime calls (e.g. for new class) don't use the callbacks anymore that dmd hooks into.
- "BumpGC" selects a GC that just forwards malloc to the bump allocator. It is not faster than the version using the GC for half the allocations, but uses about 10 GB more memory. Not sure why that is, I don't think that there are so many array operations that might reallocate instead of appending.
- cleanup:none skips the final collection before shutdown, which is probably also the longest, a simple win of up to 2% execution time

rainers committed:
remove evDone

rainers commented on druntime/src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -3600,21 +3602,34 @@ Lmark:
+            else
+                mark!(false, true, true)(ScanRange!false(pbot, ptop));
+
+            for (bool done = false; !done; )
```
It is probably not affecting GC performance much, but it is not so nice that this loop is burning CPU cycles if the scanStack is empty, but other threads are still busy. It cannot wait for evStackFilled because that might not trigger anymore or other threads might be woken up instead.

schveiguy replied:
FWIW, I have the new GC keeping scanning threads alive. I settled on 2 events, one for GC starting and one for work available. Both are manual reset.

The work available event is waited on when the work list goes to 0 but some active threads still scanning. Once both are 0, it moves to waiting on the gcstarted event.

That event is waited on until either worklist is non-zero, or it's indicated that all threads should be shut down. I denote this by setting the active threads to -1.

See the mechanisms here: https://github.com/symmetryinvestments/symgc/blob/91650bfa4fec57fafa39603273cad34da0e662ea/source/d/gc/scanner.d#L36

And the thread scan main loop here: https://github.com/symmetryinvestments/symgc/blob/91650bfa4fec57fafa39603273cad34da0e662ea/source/d/gc/scanner.d#L366

The key thing is when to set and reset the event, and when to increment/decrement active threads.

Note this is only used in windows, on linux we are using the normal GC locks. At some point I want to test switching linux to using the events.

rainers replied:
Thanks for info and link. I think the key advantage in your implementation is changing and checking the scan stack and the number of active threads under the same lock. I'll think about how that could be incorporated here, too.

Using manual events might be advantageous, too, but the threads only wait when the stack already has been seen empty. It's unlikely to have enough work items for all threads to actually do something, I wonder whether starting all waiting threads at once results in some penalty for them fighting over locks or other resources.

rainers committed:
use evDone to wake up main scan thread while it waits for other scanner threads to finish

schveiguy commented:
> * even though dmd is meant to use the bump allocator by default it still allocates half of its memory with the disabled GC. This is because the lowered druntime calls (e.g. for new class) don't use the callbacks anymore that dmd hooks into.

In a recent DLF meeting, we talked about switching the bump GC to be a full GC interface, so we don't have to hook any of these things, just change the underlying allocator. Really, a bump allocator is a GC, just garbage is collected at the end of the process.

rainers committed:
avoid having two scan stacks (for conservative or precise scanning) by counting bytes, not ScanRange entries
move popLocked() out of stack, locked pushing out of mark()
continue scanning with current region when flshsing the local stack

rainers committed:
back to manual evStackFilled event with busyThreads changed under the same lock

rainers committed:
reinit stackLock when forking

schveiguy commented:
Ah nice, I see evDone is now only waited on by the main thread. This looks much better. I'm going to give it another once-over.

schveiguy commented on druntime/src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -3600,21 +3569,25 @@ Lmark:
 
+        void pullLoop(bool precise)()
+        {
+            static if (precise)
+                mark!(true, true, true)(ScanRange!true(pbot, ptop, null));
+            else
+                mark!(false, true, true)(ScanRange!false(pbot, ptop));
```
```suggestion
            mark!(precise, true, true)(ScanRange!precise(pbot, ptop));
```
`null` is the default for precise bitmap, so we don't need this added complication.

schveiguy commented on druntime/src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -3600,21 +3569,25 @@ Lmark:
+                mark!(false, true, true)(ScanRange!false(pbot, ptop));
+
+            while (pullFromScanStack())
+                evDone.wait(1.msecs);
```
Why are we waiting for only 1ms here? Why not wait forever?

Note that the background threads are no longer waiting on this. ONLY the main thread can consume this event. So waking back up is guaranteed.

schveiguy replied:
OK, I see now, this would put the main thread to sleep until the scan stack was empty and another thread set this event.

Hard to think of a good way to have the main thread wait until the right time (more data available or all threads unbusy). In my implementation, I set both events when new data is added I think. But this might be pretty bad here.

rainers replied:
AFAICT you are also setting the `workReadyEvent` corresponding to `evStartFilled` at the end of the collection and let the threads figure out that there is nothing more to do. Doesn't that wake up all the threads again without anything to do? Probably not a big deal, though.

As evDone is set outside of the lock, there might also be some racing condition on it, but as you mentioned the main thread is the only consumer, so unlikely a problem.

rainers replied:
When waiting for the scanStack to become non-empty, a timeout of 1 ms might be quite large, so waiting for both events simultaneously might be better here. I don't think we have a portable primitive for that in core.sync.

schveiguy replied:
> AFAICT you are also setting the `workReadyEvent` corresponding to `evStartFilled` at the end of the collection and let the threads figure out that there is nothing more to do.

The two events I use mark a GC in progress (basically the same as evDone, but an event sent to the scan threads, not the main thread), and data is ready (same as stackFilled).

The reason my workReady is set when no more data is coming is so the threads can exit the wait loop for more data. At this point, the main thread exits and finishes the GC. The background threads go to sleep waiting on the next GC start.

So my event config is slightly different here. The background threads are the ones that wait on 2 different events, and the main thread only ever waits on one.

I suppose `workReady` could be something like `workReadyOrDone` maybe?

I'm also wondering how expensive a set and reset is, I probably should not call it as often...

> When waiting for the scanStack to become non-empty, a timeout of 1 ms might be quite large,

It should be fine, it just means that the main thread might not pick up as much work during a scan when getting down to the last few ranges. But when a scan is over, it should immediately wake up.

I bet it doesn't wait more than a few times per scan here.

rainers replied:
> I'm also wondering how expensive a set and reset is, I probably should not call it as often...

I have stepped through SetEvent/ResetEvent in a debugger on Windows and it goes through a number of calls with the main functionality invoked by a syscall. Sounds rather expensive expensive to me, so I tried to minimize calling them with the pre/post-empty checks on the stack.

> I bet it doesn't wait more than a few times per scan here.

When the scan stack has indeed become empty and pullFromScanStackImpl has returned, it is very likely that there are other threads waiting, too, so they can also pickup pending work.

schveiguy commented on druntime/src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -3600,21 +3569,25 @@ Lmark:
+            else
+                mark!(false, true, true)(ScanRange!false(pbot, ptop));
+
+            while (pullFromScanStack())
```
```suggestion
            while (pullFromScanStackImpl!precise())
```
Don't pay the price of checking the runtime boolean again.

rainers replied:
Good catch, updated.

schveiguy reviewed (changes requested):
Looking pretty good.

schveiguy commented on druntime/src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -3759,29 +3730,62 @@ Lmark:
+
+        // reverse order for depth-first-order traversal
+        foreach_reverse (ref range; ranges)
+            scanStack.push(range);
```
I know it's not a huge deal here, but this does a lot of checking if the stack needs expanding when we know the size we will require.

rainers replied:
I was also considering adding a primitive adding multiple ranges to avoid the duplicate "logic" of the foreach_reverse-loop for the parallel and single-threaded case. Avoiding the checks is another reason, so I added `pushReverse`.

schveiguy commented:
tracking of busy threads looks much better BTW!

rainers committed:
address review:
- add default type to ToScanStack to allow implementation of ptr and length
- simplify pullLoop

rainers commented on druntime/src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -2452,88 +2450,58 @@ struct Gcx
-        else
-            alias scanStack = toscanConservative;
-    }
+    ToScanStack scanStack;
```
The lock used to be part of ToScanStack, which complicated dealing with it from the outside with multiple instances. (In theory, you could also switch from precise to conservative after the first collection). I was also quite happy to remove a couple of conditionals with this change.

Agreed, the change to `toscanRoots` was not so nice. I now added a "default range type" to be used by `ptr()` and `length` so using `toscanRoots` doesn't need to access internals.

rainers committed:
add ToScanStack.pushReverse for pushing multiple ranges

schveiguy commented on druntime/src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -2452,88 +2450,73 @@ struct Gcx
-        else
-            alias scanStack = toscanConservative;
-    }
+    ToScanStack!void scanStack; // used with both ScanRange!false and ScanRange!true
```
I still favor a union here, and keep the list tied to a type.

It's just very easy to accidentally use the wrong thing here, and you probably won't see any errors unless there happens to be some mismatch in sizes (at runtime).

Basically, I'm thinking:
```suggestion
    union {
        ToScanStack!(ScanRange!false) scanStackConservative;
        ToScanStack!(ScanRange!true) scanStackPrecise;
    }
    template scanStack(bool precise) {
        static if(precise)
            alias scanStack = scanStackPrecise;
        else
            alias scanStack = scanStackConservative;
    }
```

And then you can use it wherever you have a `precise` compile-time bool like `scanStack!precise`

Keep the length and cap in bytes, because this will be needed if you use a different union member.

rainers replied:
Ok, done. I have restored the length to count entries in the stack to avoid having to divide by the size of an entry. Switching between conservative and precise must happen with an empty stack anyway.

rainers committed:
restore scanStackConservative/scanStackPrecise, but as a union

schveiguy committed:
Update druntime/src/core/internal/gc/impl/conservative/gc.d

Fix spacing to match original.

rainers committed:
concurrent GC: better reinit stackLock than call unlock() when forking

rainers commented on druntime/src/core/internal/gc/impl/conservative/gc.d:
```diff
@@ -3534,8 +3523,9 @@ Lmark:
                         Gcx.instance.numScanThreads = 0;
                         Gcx.instance.scanThreadData = null;
                         Gcx.instance.busyThreads = 0;
+                        Gcx.instance.stackLock.unlock();
```
The test for the concurrent GC was stalling and I could see it waiting for stackLock in the debugger. I guess forking the process with the lock held due to a spurious wakeUp. `destroy` would lose the `contention` information, though. I have changed it to reassigning the initialized value.

schveiguy commented:
thank you @rainers!

rainers commented:
Thanks @schveiguy for the reviews.
